app/src/main/java/emu/skyline/AppDialog.kt

```diff
@@ -101,6 +154,66 @@ class AppDialog : BottomSheetDialogFragment() {
             shortcutManager.requestPinShortcut(info.build(), null)
         }
 
+        val saveFolderPath = savesFolderRoot + item.titleId
+        val saveExists = File(saveFolderPath).exists()
+
+        binding.deleteSave.isEnabled = saveExists
+        binding.deleteSave.setOnClickListener {
+            AlertDialog.Builder(requireContext())
+                .setTitle(getString(R.string.delete_save_confirmation_message))
+                .setMessage(getString(R.string.action_irreversible))
+                .setPositiveButton(getString(R.string.yes)) { dialogInterface, _ ->
+                    File(saveFolderPath).deleteRecursively()
+                    binding.deleteSave.isEnabled = false
+                    binding.exportSave.isEnabled = false
+                    dialogInterface.dismiss()
+                }.setNegativeButton(getString(R.string.no)) { dialogInterface, _ ->
+                    dialogInterface.cancel()
+                }.show()
+        }
+
+        binding.importSave.setOnClickListener {
+            documentPicker.launch(arrayOf("application/zip"))
+        }
+
+        binding.exportSave.isEnabled = saveExists
+        binding.exportSave.setOnClickListener {
+            CoroutineScope(Dispatchers.IO).launch {
+                try {
+                    val saveFolder = File(saveFolderPath)
+                    val outputZipFile = File("$savesFolderRoot${item.title} (v${binding.gameVersion.text}) [${item.titleId}] - ${LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"))}.zip")
+                    lastZipCreated?.delete()
+                    outputZipFile.createNewFile()
+                    ZipOutputStream(BufferedOutputStream(FileOutputStream(outputZipFile))).use { zos ->
+                        saveFolder.walkTopDown().forEach { file ->
+                            val zipFileName = file.absolutePath.removePrefix(savesFolderRoot).removePrefix("/")
+                            if (zipFileName == "") return@forEach
+                            val entry = ZipEntry("$zipFileName${(if (file.isDirectory) "/" else "")}")
+                            zos.putNextEntry(entry)
+                            if (file.isFile) {
+                                file.inputStream().use { fis -> fis.copyTo(zos) }
+                            }
+                        }
+                    }
+                    lastZipCreated = outputZipFile
+                } catch (e: Exception) {
+                    withContext(Dispatchers.Main){
+                        Snackbar.make(binding.root, e.message as CharSequence, Snackbar.LENGTH_LONG).show()
```

Comment:
Showing Java exception messages is bad, and not useful to the user at all. Either enumerate all the possible failure points each with a proper error string, or show a generic failure message.

Reply:
Forgot to remove it. It was done to debug an issue in discord 😅